### PR TITLE
Add binder support via jupytext

### PR DESCRIPTION
**Pull request by StanczakDominik**
Add binder support via jupytext as per Matrix discussion https://matrix.to/#/!MeRdFpEonLoCwhoHeT:matrix.org/$1561981905806323ldgXZ:matrix.org?via=matrix.org&via=cadair.com&via=maunium.net

this is probably going to take me a few tries, please be patient :)

### CODE REVIEW

**Comment by StanczakDominik** [comment]
I'm getting pretty close to those binder-links-inside-sphinx-gallery stretch goals @Cadair, but I think this is all I'm going to be able to do for now without consulting with the good people upstream. Once I do, I'll do a followup PR - for now, this should be stable enough.

**Comment by StanczakDominik** [comment]
Oh yeah, please do not touch this PR without a squash. :D

File: binder/environment.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  - sphinx
+  - sphinx-astropy
+  - sphinx-gallery
+  - towncrier
+  - pip:
+    - sunpy-sphinx-theme
+    - jplephem
+    - astroquery
+    - jupytext
+    - ..
```
Comment (Cadair) [inline binder/environment.yml]: Does this want to be here?

**Comment by Cadair** [comment]
@nabobalis We don't actually need to backport this for it to be useful, unless @StanczakDominik implements the integrations with sphinx-gallery.

**Comment by nabobalis** [comment]
We won't need the config settings on the branch for this to work?

File: binder/environment.yml (hunk 1 above)
Comment (StanczakDominik) [reply binder/environment.yml]: Yes, otherwise any `import sunpy` from the Binder notebooks is going to fail.

**Comment by Cadair** [comment]
Well it depends if we want to deploy this on other branches or just master.

**Comment by StanczakDominik** [comment]
I haven't considered running this on non-master branches yet, but it would take just a simple string replace applied to the Binder link.

> @nabobalis We don't actually need to backport this for it to be useful, unless @StanczakDominik implements the integrations with sphinx-gallery.

right now I'm pretty close, I've got [this PlasmaPy branch](https://github.com/PlasmaPy/PlasmaPy/compare/master...StanczakDominik:gallery-links) and I'm trying to figure out how to monkey patch `sphinx_gallery.binder.gen_binder_rst`. I need to change

```
https://mybinder.org/v2/gh/StanczakDominik/PlasmaPy/gallery-links?filepath=plasmapy/examples/auto_examples/plot_dispersion_function.ipynb
# to
https://mybinder.org/v2/gh/StanczakDominik/PlasmaPy/gallery-links?filepath=plasmapy/examples/plot_dispersion_function.py
```

Note the removal of `auto_examples` which comes from the non-binder sphinx-gallery config and the changed format.
I'm not sure how to actually monkeypatch sphinx-gallery, though.

**Comment by StanczakDominik** [comment]
... I had a typo and I was monkeypatching `get_binder_rst` instead. Stay tuned...

**Comment by StanczakDominik** [comment]
8ed814a *should* do the trick; but I'd really appreciate tips on how to do that more cleanly.

**Comment by StanczakDominik** [comment]
No idea why that docs error pops out :|

**Comment by Cadair** [comment]
I have seen that error before, it's something to do with how it tries to copy the binder files.

I think the trick was to make `gallery_dirs` relative, [like this](https://github.com/heliophysicsPy/gallery/commit/87e515ce3d91b866048131a49146a882702d7fd8#diff-fb0d0d7ff27f33697ba3b787ec894677R219)

**Comment by Cadair** [comment]
@StanczakDominik I hate to tell you this but it [dosen't seem to have worked](https://15629-2165383-gh.circle-artifacts.com/0/root/project/docs/_build/html/generated/gallery/acquiring_data/downloading_hmi.html#sphx-glr-generated-gallery-acquiring-data-downloading-hmi-py)

**Comment by StanczakDominik** [comment]
Yeah, I see. I'll come back to this tomorrow.

**Comment by Cadair** [comment]
As mentioned in matrix, anything done without a patch upstream is going to be a hack lol.

**Comment by StanczakDominik** [comment]
I'll try to discuss adding support for this kind of custom link wizardry
with the gallery folks.

>

File: binder/environment.yml (hunk 1 above)
Comment (nabobalis) [inline binder/environment.yml]: What is this?

Comment (Cadair) [reply binder/environment.yml]: It installs the git repo in the directory above. Given I asked the [same question](https://github.com/sunpy/sunpy/pull/3256#discussion_r299040618) we should probably comment this.

**Comment by Cadair** [comment]
@StanczakDominik can you merge or rebase to get the CI fixes from #3257 please?

**Comment by StanczakDominik** [comment]
:green_heart:

**Pull request merged**